src/lib/fastDownward/analysis.ts

```diff
@@ -0,0 +1,207 @@
+import { LearningUnit, Path, Skill, SkillAnalyzedPath } from "../types";
+import { SearchNode } from "./searchNode";
+import { State } from "./state";
+import { GlobalKnowledge } from "./global-knowledge";
+
+/**
+ * Compute which LearningUnits are reversed reachable based on the given goals in state.
+ * Using teachingGoals of the LearningUnits and the skill groups
+ */
+function availableActions<LU extends LearningUnit>(
+	currentState: State,
+	learningUnits: ReadonlyArray<LU>,
+	globalKnowledge: GlobalKnowledge
+) {
+    // Find the LearningUnits for each goal in the state using teachingGoals.
+    let usefulLus = learningUnits.filter(lu => lu.teachingGoals.some(skill => currentState.learnedSkills.includes(skill.id)));
+    
+    // Find the LearningUnits for each goal in the state using globalKnowledge (Skill groups).
+    globalKnowledge.getAllParents().forEach(parent => {
+        if (currentState.getHashCode().includes(parent.id) ) {
+            const lu = learningUnits.filter(lu => lu.teachingGoals.some(skill => parent.nestedSkills.includes(skill.id)));
+            usefulLus = usefulLus.concat(lu);
+        }
+    });
+    
+    // Remove duplicate LearningUnits
+    return [...new Set(usefulLus)];
+}
+
+/**
+ * Analysis a goal to find the missing skills.
+ * @param goal The skills that should be learned (should not contain skills that are already known).
+ * @param globalKnowledge The set of all skills (independent of what was already learned and what should be learned).
+ * @param learningUnits All learning units of the system to learn new skills.
+ * @returns A list of the missing skills with the sub paths for them. 
+*/
+export function skillAnalysis<LU extends LearningUnit>(
+    globalKnowledge: GlobalKnowledge,
+    learningUnits: ReadonlyArray<LU>,
+    goal: Skill[]
+): SkillAnalyzedPath[] | null {
+
+    const openListMap = new Map();
+    let currentNode: SearchNode<LU>;
+
+    // Create initial state with goals and globalKnowledge
+    const initialState = new State(
+        goal.map(skill => skill.id),
+        globalKnowledge
+    );
+
+    // Remove duplicate LearningUnits
+    initialState.learnedSkills = [...new Set(initialState.learnedSkills)];
+
+    const openList: SearchNode<LU>[] = [new SearchNode<LU>(initialState, null, null, 0, 0)];
+    
+    // Use goalString to find missing skills that does not have a requirement
+    let goalString = `,`.concat(initialState.learnedSkills.join(",").concat(`,`));

```

Comment:
* Please explain idea of goalString (seems a very importan concept of the algorithm)
* Is a String + many concat operations the best approach or would be an array a better solution?

Reply:
- goalString is a string with all the skills for the request goal (or skill) to analyze.

Case 1: If a skill does not have requirements, then we can't find it through skill requirements tracing. 
Case 2: If a skill depends on skill groups (parent/child), then we can't find it through skill requirements tracing.
We use goalString to find the missing skills that do not have requirements and the skill that uses skill groups.

- We use String since it is easy and quick to find and remove skills (better than array). 
- Concat to avoid finding numbers that have a similar number (as string) like "10", is shared between 100, 1005, and 1105. So, we concat commas around it to ",10,".